gitlab.go

```diff
@@ -0,0 +1,195 @@
+package reviewdog
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/xanzy/go-gitlab"
+
+	"golang.org/x/sync/errgroup"
+)
+
+var _ CommentService = &GitLabMergeRequest{}
+var _ DiffService = &GitLabMergeRequest{}
+
+// GitLabMergeRequest is a comment and diff service for GitLab MergeRequest.
+//
+// API:
+//
+//
+type GitLabMergeRequest struct {
+	cli      *gitlab.Client
+	owner    string
+	repo     string
+	pr       int
+	sha      string
+	projects string
+
+	muComments   sync.Mutex
+	postComments []*Comment
+
+	postedcs postedcomments
+
+	// wd is working directory relative to root of repository.
+	wd string
+}
+
+// NewGitLabMergeReqest returns a new GitLabMergeRequest service.
+// GitLabMergeRequest service needs git command in $PATH.
+func NewGitLabMergeReqest(cli *gitlab.Client, owner, repo string, pr int, sha string) (*GitLabMergeRequest, error) {
+	workDir, err := gitRelWorkdir()
+	if err != nil {
+		return nil, fmt.Errorf("GitLabMergeRequest needs 'git' command: %v", err)
+	}
+	return &GitLabMergeRequest{
+		cli:      cli,
+		owner:    owner,
+		repo:     repo,
+		pr:       pr,
+		sha:      sha,
+		projects: owner + "/" + repo,
+		wd:       workDir,
+	}, nil
+}
+
+// Post accepts a comment and holds it. Flush method actually posts comments to
+// GitLab in parallel.
+func (g *GitLabMergeRequest) Post(_ context.Context, c *Comment) error {
+	c.Path = filepath.Join(g.wd, c.Path)
+	g.muComments.Lock()
+	defer g.muComments.Unlock()
+	g.postComments = append(g.postComments, c)
+	return nil
+}
+
+// Flush posts comments which has not been posted yet.
+func (g *GitLabMergeRequest) Flush(ctx context.Context) error {
+	defer g.muComments.Unlock()
+	g.muComments.Lock()
+
+	if err := g.setPostedComment(ctx); err != nil {
+		return err
+	}
+
+	return g.postCommentsForEach(ctx)
+}
+
+func (g *GitLabMergeRequest) postCommentsForEach(ctx context.Context) error {
+	var eg errgroup.Group
+	for _, c := range g.postComments {
+		comment := c
+		if g.postedcs.IsPosted(comment, comment.Lnum) {
+			continue
+		}
+		eg.Go(func() error {
+			commitID, err := g.getLastCommitsID(comment.Path, comment.Lnum)
+			if err != nil {
+				return err
+			}
+			body := commentBody(comment)
+			ltype := "new"
+			prcomment := &gitlab.PostCommitCommentOptions{
+				Note:     &body,
+				Path:     &comment.Path,
+				Line:     &comment.Lnum,
+				LineType: &ltype,
+			}
+			_, _, err = g.cli.Commits.PostCommitComment(g.projects, commitID, prcomment, nil)
+			return err
+		})
+	}
+	return eg.Wait()
+}
+
+func (g *GitLabMergeRequest) getLastCommitsID(path string, line int) (string, error) {
+	lineFormat := fmt.Sprintf("%d,%d", line, line)
+	s, err := exec.Command("git", "blame", "-l", "-L", lineFormat, path).Output()
+	if err != nil {
+		return "", fmt.Errorf("failed to get commitID: %v", err)
+	}
+	commitID := strings.Split(string(s), " ")[0]
```

Comment:
Can you add check to ensure we can access 0 idx for safety?

Reply:
My understanding `strings.Split` return at least 1 item.
should I check the length of items?

https://golang.org/pkg/strings/#Split

Reply:
oh, that's true. never mind.